codegen/src/api/mod.rs

```diff
@@ -378,84 +342,91 @@ where
-            let struct_def = CompositeDef::struct_def(
-                &ty,
-                struct_name.as_ref(),
-                Default::default(),
-                fields.clone(),
-                Some(parse_quote!(pub)),
-                type_gen,
-                docs,
-                crate_path,
-                Some(alias_module_name.clone()),
-            )?;
-
-            let type_aliases = TypeAliases::new(fields, types_mod_ident.clone(), alias_module_name);
-
-            Ok((var.name.to_string(), struct_def, type_aliases))
+            let mut composite = CompositeIR::new(
+                syn::parse_str(&struct_name).expect("enum variant name is valid ident"),
+                composite_ir_kind,
+                type_gen.docs_from_scale_info(&var.docs),
+            );
+
+            let type_alias_mod = generate_type_alias_mod(&mut composite, type_gen);
+            Ok(StructFromVariant {
+                variant_name: var.name.to_string(),
+                composite,
+                type_alias_mod,
+            })
         })
         .collect()
 }
 
-/// Generate the type aliases from a set of enum / struct definitions.
-///
-/// The type aliases are used to make the generated code more readable.
-#[derive(Debug)]
-pub struct TypeAliases {
-    fields: CompositeDefFields,
-    types_mod_ident: syn::Ident,
-    mod_name: syn::Ident,
-}
-
-impl TypeAliases {
-    pub fn new(
-        fields: CompositeDefFields,
-        types_mod_ident: syn::Ident,
-        mod_name: syn::Ident,
-    ) -> Self {
-        TypeAliases {
-            fields,
-            types_mod_ident,
-            mod_name,
-        }
-    }
+pub struct StructFromVariant {
+    variant_name: String,
+    composite: CompositeIR,
+    type_alias_mod: TokenStream2,
 }
 
-impl quote::ToTokens for TypeAliases {
-    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
-        let has_fields = matches!(&self.fields, CompositeDefFields::Named(fields) if !fields.is_empty())
-            || matches!(&self.fields, CompositeDefFields::Unnamed(fields) if !fields.is_empty());
-        if !has_fields {
-            return;
-        }
-
-        let visibility: syn::Visibility = parse_quote!(pub);
-
-        let aliases = self
-            .fields
-            .to_type_aliases_tokens(Some(visibility).as_ref());
+/// Modifies the composite, by replacing its types with references to the generated type alias module.
+/// Returns the TokenStream of the type alias module.
+///
+/// E.g a struct like this:
+/// ```ignore
+/// pub struct SetMaxCodeSize {
+///     pub new: ::core::primitive::u32,
+/// }
+/// ```
+/// will be made into this:
+/// ```ignore
+/// pub struct SetMaxCodeSize {
+///     pub new: set_max_code_size::New,
+/// }
+/// ```
+/// And the type alias module will look like this:
+/// ```ignore
+/// pub mod set_max_code_size {
+///     use super::runtime_types;
+///     pub type New = ::core::primitive::u32;
+/// }
+/// ```
+pub fn generate_type_alias_mod(
```

Comment:
Is this used for type substitution or when is it used?

Reply:
This is used in the `generate_structs_from_variants` function above that we use to generate special functions for `events` and `calls`. For each of these calls/events we also generate a struct that bundles together the arguments needed for the call / event query. Because some of these arguments are very long complicated types, we do not use the types directly, instead we generate a small module next to this the struct for each call / event. The module has the same name as the struct, just in snake case. It contains a type alias for each field of the struct. This `generate_type_alias_mod` function is responsible for generating these little modules for each call / event. 
But it does not change much in terms of usage, the type aliases are basically just syntactic sugar. No types are actually changed/substituted.

Reply:
got it, thanks